src/primitives/transaction.h

```diff
@@ -351,11 +356,36 @@ class CTransaction
 
     template <typename Stream, typename Operation>
     inline void SerializationOp(Stream& s, Operation ser_action, int nType, int nVersion) {
-        READWRITE(*const_cast<int32_t*>(&this->nVersion));
+        if (ser_action.ForRead()) {
+            // When deserializing, unpack the 4 byte header to extract fOverwintered and nVersion.
+            uint32_t header;
+            READWRITE(header);
+            *const_cast<bool*>(&fOverwintered) = header >> 31;
+            *const_cast<int32_t*>(&this->nVersion) = header & 0x7FFFFFFF;
+        } else {
+            if (fOverwintered) {
+                // When serializing v3, assemble the 4 byte header from fOverwintered and nVersion
+                uint32_t header = (fOverwintered << 31) | this->nVersion;
```

Comment:
I think we can simplify the code and just use this in both cases, since `fOverwintered << 31` will be zero if `fOverwintered == false`. @daira?

Reply:
It could be 1 << 31 or 0x80000000

Reply:
Right; what I mean is that if `fOverwintered == false`, then `header == this->nVersion`, so we can get rid of the `if (fOverwintered)` and the other branch.

Reply:
Simplified.